runtime(zip): support PowerShell Core

ShayHill opened the pull request:
related: #17987

If `g:zip_[zip|unzip|extract]cmd` is not executable or fails, try PowerShell core `if &shell =~ 'pwsh'`.

If the gnu executable (e.g., g:zip_zipcmd) does not exist or if the gnu command fails for any reason, fallback to a PowerShell Core command `if &shell =~ 'pwsh'`. If none of these paths succeed:

- exe exists -> gnu_cmd_succeeds
- exe exists -> gnu_cmd_fails -> ps_cmd_succeeds
- exe does not exist -> ps_cmd_succeeds
... echomsg all failures.

To avoid any potential for failures for non-pwsh users, this pr will not even create a PowerShell Core command unless &shell =~ 'pwsh'.

The 'Path Traversal Attack handling' triggers on this criterion:

`if fname =~ '^[.]\{1,2}/'`

Everything works if I test with a different criterion, but I cannot test *that* exact criterion, because no such filename can be created in Windows (AFAICT).

PowerShell cannot write or extract a file from within a subdirectory. In fact, Windows in general cannot, at least not in a straightforward way. It wouldn't be impossible—I'm sure PowerShell scripting language is Turing complete—but we'd have to make an arbitrary decision about a name for the extracted file. Windows filenames cannot have slashes. I suppose we could use `%2f` if we ever wanted to build that.

For now, this pr issues a warning if a user reads a file in a subdirectory of an archive, letting the user know that they won't be able to write any changes.

ShayHill committed:
runtime(zip): support PowerShell Core

related: #17987

If `g:zip_[zip|unzip|extract]cmd` is not executable or fails, try
PowerShell core `if &shell =~ 'pwsh'`.

ShayHill committed:
runtime(zip): remove leading l: for local vars

ShayHill committed:
runtime(zip): document PowerShell Core support

chrisbra commented on runtime/autoload/zip.vim:
```diff
@@ -78,15 +79,124 @@ if v:version < 901
+  else
+    call add(failures, a:executable.' not available on your system')
+  endif
+  if a:0 == 1
```
you may want to use here:

```suggestion
  if &shell =~ 'pwsh' && a:0 == 1
```

chrisbra commented on runtime/autoload/zip.vim:
```diff
@@ -140,7 +250,14 @@ fun! zip#Browse(zipfile)
+    call s:TryExecGnuFallBackToPs(g:zip_unzipcmd, gnu_cmd, ps_cmd)
+  else
+    call s:TryExecGnuFallBackToPs(g:zip_unzipcmd, gnu_cmd)
+  endif
```
Then you could chnage this part here to:
```vim
  let gnu_cmd = "keepj sil r! " . g:zip_unzipcmd . " -Z1 -- " . s:Escape(a:zipfile, 1)
  let ps_cmd = 'keepj sil r! ' . s:ZipBrowsePS(a:zipfile)
  call s:TryExecGnuFallBackToPs(g:zip_unzipcmd, gnu_cmd, ps_cmd)
```

ShayHill replied:
I welcome this switch because it much cleaner, but it leaves a possibility for one of the `s:[something]PS` functions to fail and spoil things for non-PowerShell users. If none of the tests fail, then I suppose that is a remote possibility.

chrisbra commented on runtime/autoload/zip.vim:
```diff
@@ -220,7 +337,15 @@ fun! zip#Read(fname,mode)
+    call s:TryExecGnuFallBackToPs(g:zip_unzipcmd, gnu_cmd, ps_cmd)
+  else
+    call s:TryExecGnuFallBackToPs(g:zip_unzipcmd, gnu_cmd)
+  endif
```
same here

chrisbra commented on runtime/autoload/zip.vim:
```diff
@@ -273,7 +398,14 @@ fun! zip#Write(fname)
+      call s:TryExecGnuFallBackToPs(g:zip_zipcmd, gnu_cmd, ps_cmd)
+    else
+      call s:TryExecGnuFallBackToPs(g:zip_zipcmd, gnu_cmd)
+    endif
```
and here

chrisbra commented on runtime/autoload/zip.vim:
```diff
@@ -299,7 +431,25 @@ fun! zip#Write(fname)
+    let ps_cmd = s:ZipUpdatePS(s:Escape(fnamemodify(zipfile, ':p'), 0), s:Escape(fname, 0))
+    let ps_cmd = 'call system(''' . substitute(ps_cmd, "'", "''", 'g') . ''')'
+    call LogMessage("PS Command: " . ps_cmd)
+    call LogMessage("GNU Command: " . gnu_cmd)
```
LogMessage seems undefined?

chrisbra commented on runtime/doc/pi_zip.txt:
```diff
@@ -77,6 +77,14 @@ Copyright: Copyright (C) 2005-2015 Charles E Campbell	 *zip-copyright*
+   FALLBACK TO POWERSHELL CORE~
+
+   This plugin will attempt to use the (more capable) GNU zip/unzip commands
+   first. If these commands fail and if the user has set shell =~ 'pwsh', the
```
```suggestion
   first. If these commands fail but the user is using PowerShell Core (that is 'shell' option matches "pwsh"), the
```

ShayHill committed:
runtime(zip): condition fallback to PS on shell =~ 'pwsh'

Within TryExecGnuFallBackToPs, condition the PowerShell fallback to [ps function provided] && shell =~ 'pwsh'

Co-authored-by: Christian Brabandt <cb@256bit.org>

ShayHill committed:
runtime(zip): remove errant LogMessage calls

Co-authored-by: Christian Brabandt <cb@256bit.org>

ShayHill committed:
runtime(zip): clarify conditions for PS fallback

ShayHill committed:
runtime(zip): handle more PS fallback logic in function.

Function TryExecGnuFallBackToPs now handles more of the logic for when a
failed GNU call will trigger a call to a fallback PowerShell Core cmdlet
or script.

ShayHill committed:
runtime(zip): fix double call to TryExec...
